mllib/src/test/scala/org/apache/spark/ml/image/ImageSchemaSuite.scala

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.image
+
+import java.nio.file.Paths
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.ml.image.ImageSchema._
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+
+class ImageSchemaSuite extends SparkFunSuite with MLlibTestSparkContext {
+  // Single column of images named "image"
+  private val imageDFSchema =
+    StructType(StructField("image", ImageSchema.columnSchema, true) :: Nil)
+  private lazy val imagePath =
+    Thread.currentThread().getContextClassLoader.getResource("test-data/images").getPath
+
+  test("Smoke test: create basic ImageSchema dataframe") {
+    val origin = "path"
+    val width = 1
+    val height = 1
+    val nChannels = 3
+    val data = Array[Byte](0, 0, 0)
+    val mode = "CV_8UC3"
+
+    // Internal Row corresponds to image StructType
+    val rows = Seq(Row(Row(origin, height, width, nChannels, mode, data)),
+      Row(Row(null, height, width, nChannels, mode, data)))
+    val rdd = sc.makeRDD(rows)
+    val df = spark.createDataFrame(rdd, imageDFSchema)
+
+    assert(df.count == 2, "incorrect image count")
+    assert(ImageSchema.isImageColumn(df, "image"), "data do not fit ImageSchema")
+  }
+
+  test("readImages count test") {
+    var df = readImages(imagePath, recursive = false)
+    assert(df.count == 0)
+
+    df = readImages(imagePath, recursive = true, dropImageFailures = false)
+    assert(df.count == 8)
+
+    df = readImages(imagePath, recursive = true, dropImageFailures = true)
+    val count100 = df.count
+    assert(count100 == 7)
+
+    df = readImages(imagePath, recursive = true, sampleRatio = 0.5, dropImageFailures = true)
+    // Random number about half of the size of the original dataset
+    val count50 = df.count
+    assert(count50 > 0.2 * count100 && count50 < 0.8 * count100)
+  }
+
+  test("readImages partition test") {
+    val df = readImages(imagePath, recursive = true, dropImageFailures = true, numPartitions = 3)
+    assert(df.rdd.getNumPartitions == 3)
+  }
+
+  // Images with the different number of channels
+  test("readImages pixel values test") {
+
+    val images = readImages(imagePath + "/multi-channel/", recursive = false).collect
+
+    images.foreach{
+      rrow => {
+        val row = rrow.getAs[Row](0)
+        val filename = Paths.get(getOrigin(row)).getFileName().toString()
+        if(firstBytes20.contains(filename)) {
+          val mode = getMode(row)
+          val bytes20 = getData(row).slice(0, 20)
+
+          val expectedMode = firstBytes20(filename)._1
+          val expectedBytes = firstBytes20(filename)._2
+
+          assert(expectedMode == mode, "mode of the image is not read correctly")
+
+          if (!compareBytes(expectedBytes, bytes20)) {
+            throw new Exception("incorrect numeric value for flattened image")
```

Comment:
Assert this rather than check it and manually throw an exception

Reply:
done